Frameworks/ImageIO/CGImageSource.mm

```diff
@@ -25,50 +28,327 @@
+const CFStringRef kUTTypeJPEG = static_cast<const CFStringRef>(@"public.jpeg");
+const CFStringRef kUTTypeTIFF = static_cast<const CFStringRef>(@"public.tiff");
+const CFStringRef kUTTypeGIF = static_cast<const CFStringRef>(@"com.compuserve.gif");
+const CFStringRef kUTTypePNG = static_cast<const CFStringRef>(@"public.png");
+const CFStringRef kUTTypeBMP = static_cast<const CFStringRef>(@"com.microsoft.bmp");
+const CFStringRef kUTTypeICO = static_cast<const CFStringRef>(@"com.microsoft.ico");
+
+@implementation ImageSource
+- (instancetype)initWithData:(CFDataRef)data {
+    if (self = [super init]) {
+        _data = (NSData*)data;
+    }
+
+    return self;
+}
+
+- (instancetype)initWithURL:(CFURLRef)url {
+    if (self = [super init]) {
+        _data = [NSData dataWithContentsOfURL:(NSURL*)url];
+    }
+
+    return self;
+}
+
+- (instancetype)initWithDataProvider:(CGDataProviderRef)provider {
+    if (self = [super init]) {
+        _data = (NSData*)CGDataProviderCopyData(provider);
+    }
+
+    return self;                               
+}
+
+// Helper function to identify image format from image byte stream
+- (CFStringRef)getImageType {
+    char imageIdentifier[12] = {0};
+    [self.data getBytes:&imageIdentifier length:12];
+    static const unsigned char BMPIdentifier[] = {'B','M'};
+    static const unsigned char GIFIdentifier[] = {'G','I','F'};
+    static const unsigned char ICOIdentifier[] = {0x00, 0x00, 0x01, 0x00};
+    static const unsigned char JPEGIdentifier[] = {0xFF, 0xD8, 0xFF};
+    static const unsigned char PNGIdentifier[] = {0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};
+    static const unsigned char TIFFIdentifier1[] = {'M', 'M', 0x00, 0x2A};
+    static const unsigned char TIFFIdentifier2[] = {'I', 'I', 0x2A, 0x00};
+    CFStringRef imageFormat;
+
+    if (!memcmp(imageIdentifier, BMPIdentifier, 2)) {
+        imageFormat = kUTTypeBMP;
+    } else if (!memcmp(imageIdentifier, GIFIdentifier, 3)) {
+        imageFormat = kUTTypeGIF;
+    } else if (!memcmp(imageIdentifier, ICOIdentifier, 4)) {
+        imageFormat = kUTTypeICO;
+    } else if (!memcmp(imageIdentifier, JPEGIdentifier, 3)) {
+        imageFormat = kUTTypeJPEG;
+    } else if (!memcmp(imageIdentifier, PNGIdentifier, 8)) {
+        imageFormat = kUTTypePNG;
+    } else if (!memcmp(imageIdentifier, TIFFIdentifier1, 4) || !memcmp(imageIdentifier, TIFFIdentifier2, 4)) {
+        imageFormat = kUTTypeTIFF;
+    } else {
+        imageFormat = nullptr;
+    }
+
+    if (!imageFormat) {
+        UNIMPLEMENTED_WITH_MSG("Image format is not supported. "
+                               "Current release supports JPEG, BMP, PNG, GIF, TIFF & ICO image formats only.");
+    } 
+
+    return imageFormat;
+}
+@end
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes The current implementation supports common image file formats such as JPEG, GIF, TIFF, BMP, PNG and ICO. 
```

Comment:
nit: caveat should also include lack of support for kCGImageSourceTypeIdentifierHint